4.gene-expression-signatures/scripts/1.optimize_models_latent_dimensions.py

```diff
@@ -0,0 +1,122 @@
+model_names = ["pca", "ica", "nmf", "vanillavae", "betavae", "betatcvae"]
+
+for num_components in latent_dims:
+    for model_name in model_names:
+        model = None  # Initialize model as None for each iteration
+        
+        if model_name in ["pca", "ica", "nmf"]:
+            if model_name == "pca":
+                model = PCA(n_components=num_components)
+            elif model_name == "ica":
+                model = FastICA(n_components=num_components)
+            elif model_name == "nmf":
+                model = NMF(n_components=num_components, init='nndsvd', max_iter=2000, random_state=0)
+            
+            # Fit model to data
+            model.fit(train_data)
+        
+        elif model_name == "betavae":
+            study = optuna.create_study(direction="minimize")
+            study.optimize(lambda trial: objective(trial, train_tensor, train_tensor, train_data, latent_dim=num_components), n_trials=50)
+            best_trial = study.best_trial
+            model = BetaVAE(input_dim=train_data.shape[1], latent_dim=num_components, beta=best_trial.params['beta'])
+            train_loader = DataLoader(TensorDataset(train_tensor), batch_size=best_trial.params['batch_size'], shuffle=True)
+            optimizer = get_optimizer(best_trial.params['optimizer_type'], model.parameters(), best_trial.params['learning_rate'])
+            train_vae(model, train_loader, optimizer, best_trial.params['epochs'])
+        
+        elif model_name == "betatcvae":
+            study = optuna.create_study(direction="minimize")
+            study.optimize(lambda trial: objective_tc(trial, train_tensor, train_tensor, train_data, latent_dim=num_components), n_trials=50)
+            best_trial = study.best_trial
+            model = BetaTCVAE(input_dim=train_data.shape[1], latent_dim=num_components, beta=best_trial.params['beta'])
+            train_loader = DataLoader(TensorDataset(train_tensor), batch_size=best_trial.params['batch_size'], shuffle=True)
+            optimizer = get_optimizer_tc(best_trial.params['optimizer_type'], model.parameters(), best_trial.params['learning_rate'])
+            train_tc_vae(model, train_loader, optimizer, best_trial.params['epochs'])
+
+        elif model_name == "vanillavae":
+            study = optuna.create_study(direction="minimize")
+            study.optimize(lambda trial: objective_vvae(trial, train_tensor, train_tensor, train_data, latent_dim=num_components), n_trials=50)
+            
+            best_trial = study.best_trial
+            model = VanillaVAE(input_dim=train_data.shape[1], latent_dim=num_components)
+            train_loader = DataLoader(TensorDataset(train_tensor), batch_size=best_trial.params['batch_size'], shuffle=True)
+            optimizer = get_optimizer_vvae(best_trial.params['optimizer_type'], model.parameters(), best_trial.params['learning_rate'])
+            train_vvae(model, train_loader, optimizer, best_trial.params['epochs'])
```

Comment:
Consider adding comments throughout this block to help describe what's happening and why.

Reply:
This will all be updated and included in the next PR